### DISCUSSION TITLE
Format form errors for API

### DISCUSSION THREAD

**Discussion by arirangz** [symfony/symfony#44661] [Q&A] [answered]
Hello,

I am using Symfony 5 as an API. To keep to power of validation, I am still using symfony forms in my controller.

To get the errors in a readable array I have to do this in each controller:

```
$formErrors = $form->getErrors(true, true);
$errorsArray = [];
foreach ($formErrors as $error) {
    $errorsArray[$error->getOrigin()->getName()][] = $error->getMessage();
}
```

What would be the best practice to refactor this code somewhere else?

Thanks!

**Comment by lyrixx** [comment, marked as answer]
The form component is not really suited for API (it does not handle correctly boolean value)
It's better to use the serializer component for that.

You have an exemple here https://speakerdeck.com/lyrixx/symfony-live-2017-serializer?slide=62 (it's a bit outdated for the DIC part, it's better to inject the services you need instead)

**Reply by arirangz** [reply]
Thanks for your help!

I think it would be quite a lot of work to change my code not using forms...

What is the issue with forms and boolean? This issue is still in Symfony 5?

Thanks

**Reply by stof** [reply]
Well, for instance, using a CheckboxType to handle a boolean field in your API will be quite confusing, because HTML checkboxes don't work by submitting a different value. Unchecked checboxes are represented by the _absence_ of submitted values.

**Reply by arirangz** [reply]
Ok I see. We don't have this issue as our front is made with Vue.js so we are sending correct values to the api even for checkox.

**Reply by arirangz** [reply]
I read that Normalizers could help me to avoid my code in each controller but I am not sure if it's the right direction.

**Reply by lyrixx** [reply]
Yes this is the right direction 👍🏼

**Reply by arirangz** [reply]
@lyrixx Ok thanks!
Do you have any link to a tutorial on how to use Normalizers to modify getErrors forms?
Thanks

**Reply by lyrixx** [reply]
Look at my slides :) you have eveything in it. otherwise the documentation.

But basically create a new normalizer that support `FormErrorIterator` and move the code you have in your controller to it.

Then in you controller:

```php
return $this->json($form->getErrors());
```
